ur_robot_driver/src/hardware_interface.cpp

```diff
@@ -25,26 +25,429 @@
+        std::move(tool_comm_setup), calibration_checksum, (uint32_t)reverse_port, (uint32_t)script_sender_port,
+        servoj_gain, servoj_lookahead_time, non_blocking_read_);
+  }
+  catch (urcl::ToolCommNotAvailable& e)
+  {
+    RCLCPP_FATAL_STREAM(rclcpp::get_logger("URPositionHardwareInterface"), "See parameter use_tool_communication");
+
+    return return_type::ERROR;
+  }
+  catch (urcl::UrException& e)
+  {
+    RCLCPP_FATAL_STREAM(rclcpp::get_logger("URPositionHardwareInterface"), e.what());
+    return return_type::ERROR;
+  }
+
+  ur_driver_->startRTDECommunication();
+
+  status_ = hardware_interface::status::STARTED;
+
+  RCLCPP_INFO(rclcpp::get_logger("URPositionHardwareInterface"), "System successfully started!");
+
+  return return_type::OK;
+}
+
+return_type URPositionHardwareInterface::stop()
+{
+  RCLCPP_INFO(rclcpp::get_logger("URPositionHardwareInterface"), "Stopping ...please wait...");
+
+  std::this_thread::sleep_for(std::chrono::seconds(2));
+
+  ur_driver_.reset();
+
+  status_ = hardware_interface::status::STOPPED;
+
+  RCLCPP_INFO(rclcpp::get_logger("URPositionHardwareInterface"), "System successfully stopped!");
+
+  return return_type::OK;
+}
+
+template <typename T>
+void URPositionHardwareInterface::readData(const std::unique_ptr<rtde::DataPackage>& data_pkg,
+                                           const std::string& var_name, T& data)
+{
+  if (!data_pkg->getData(var_name, data))
+  {
+    // This throwing should never happen unless misconfigured
+    std::string error_msg = "Did not find '" + var_name + "' in data sent from robot. This should not happen!";
+    throw std::runtime_error(error_msg);
+  }
+}
+
+template <typename T, size_t N>
+void URPositionHardwareInterface::readBitsetData(const std::unique_ptr<rtde::DataPackage>& data_pkg,
+                                                 const std::string& var_name, std::bitset<N>& data)
+{
+  if (!data_pkg->getData<T, N>(var_name, data))
+  {
+    // This throwing should never happen unless misconfigured
+    std::string error_msg = "Did not find '" + var_name + "' in data sent from robot. This should not happen!";
+    throw std::runtime_error(error_msg);
+  }
+}
+
+return_type URPositionHardwareInterface::read()
+{
+  RCLCPP_INFO(rclcpp::get_logger("URPositionHardwareInterface"), "Reading ...");
+
+  std::unique_ptr<rtde::DataPackage> data_pkg = ur_driver_->getDataPackage();
+
+  if (data_pkg)
+  {
+    packet_read_ = true;
+    readData(data_pkg, "actual_q", urcl_joint_positions_);
+    readData(data_pkg, "actual_qd", urcl_joint_velocities_);
+    readData(data_pkg, "actual_current", urcl_joint_efforts_);
+
+    memcpy(&position_states_[0], &urcl_joint_positions_[0], 6 * sizeof(double));
+    memcpy(&velocity_states_[0], &urcl_joint_velocities_[0], 6 * sizeof(double));
+    memcpy(&joint_efforts_[0], &urcl_joint_efforts_[0], 6 * sizeof(double));
```

Comment:
memcpy makes me a little nervous. I just don't see it used very often.

Why is it needed here but not in the [ROS1 version](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/blob/master/ur_robot_driver/src/hardware_interface.cpp#L430)?

Reply:
I will use the `urcl_<>` data storage everywhere.